karabo_bridge/simulation.py

```diff
@@ -29,125 +29,183 @@
+
+    def __init__(self, source='', corr=False, gen='random'):
+        self.source = source or 'INST_DET_GENERIC/DET/detector'
+        self.corrected = corr
+        if gen == 'random':
+            self.genfunc = self.random
+        elif gen == 'zeros':
+            self.genfunc = self.zeros
+        else:
+            raise NotImplementedError('gen func %r not implemented' % gen)
+
+    @property
+    def data_shape(self):
+        return (self.modules, self.mod_y, self.mod_x, self.pulses)
 
-    pulse_count = detector_info['pulses']
-    array_shape = tuple((detector_info['modules'], ) +
-                        detector_info['module_shape'] + (pulse_count, ))
+    @property
+    def data_type(self):
+        return np.float32 if self.corrected else np.uint16
 
-    # detector random data
-    if corrected:
-        gain_data = np.zeros(array_shape, dtype=np.uint16)
-        domain = detector_info['source'].partition('/')[0]
+    def corr_passport(self):
+        domain = self.source.partition('/')[0]
         passport = [
             '%s/CAL/THRESHOLDING_Q1M1' % domain,
             '%s/CAL/OFFSET_CORR_Q1M1' % domain,
             '%s/CAL/RELGAIN_CORR_Q1M1' % domain
         ]
+        return passport
+
+    def random(self):
+        return np.random.uniform(low=1500, high=1600,
+                                 size=self.data_shape).astype(self.data_type)
+
+    def zeros(self):
+        return np.zeros(self.data_shape, dtype=self.data_type)
+
+    def module_position(self, ix):
+        y, x = np.where(self.layout == ix)
+        assert len(y) == len(x) == 1
+        return x[0], y[0]
+
+    @staticmethod
+    def gen_metadata(source, timestamp, trainId):
+        sec, frac = str(timestamp).split('.')
+        meta = {}
+        meta[source] = {
+            'source': source,
+            'timestamp': timestamp,
+            'timestamp.tid': trainId,
+            'timestamp.sec': sec,
+            'timestamp.frac': frac.ljust(18, '0')  # attosecond resolution
+        }
+        return meta
+
+    def gen_data(self, trainId):
+        data = dict(self.const_fields)
+
+        timestamp = time()
+        img = self.genfunc()
+        base_src = '/'.join((self.source.rpartition('/')[0], '{}CH0:xtdf'))
+        sources = [base_src.format(i) for i in range(16)]
+
+        data['__intime'] = timestamp
+        # TODO: cellId differ between AGIPD/LPD
+        data['cellId'] = np.arange(self.pulses, dtype=np.uint16)
+        data['combinedFrom'] = sources
+        data['image.data'] = img
+        data['length'] = np.full((self.pulses, 1), img.nbytes, dtype=np.uint32)
+        data['modulesPresents'] = [True for i in range(self.modules)]
+        if self.corrected:
+            data['image.gain'] = np.zeros(self.data_shape, dtype=np.uint16)
+            data['passport'] = self.corr_passport()
+        data['pulseCount'] = self.pulses
+        # TODO: pulseId differ between AGIPD/LPD
```

Comment:
I don't think I was aware of this. How do they differ?

Reply:
I need to recheck but I think, at least for the last experiment at SPB, `pulseId` values are different,